topics-real.Rmd

```diff
@@ -0,0 +1,55 @@
+#  (PART) Advanced Topics {-}
+
+# Making real requests {#real-requests-chapter}
+
+As touched upon in the Whole Games section, it's good to have _some_ tests against the real API.
+Indeed, the web resource can change.
+
+## What can change?
+
+What can happen?
+
+* An API introducing rate-limiting;
+* A web resource disappearing;
+* etc.
+
+## How to make real requests
+
+Maybe you can just run the same tests without using the mock files.
+
+* with vcr, this behavior is [one environment variable away](https://docs.ropensci.org/vcr/reference/lightswitch.html) (namely, `VCR_TURN_OFF`).
+* with httptest you can create the [same kind of behavior](https://enpiar.com/r/httptest/index.html#how-do-i-switch-between-mocking-and-real-requests).
+* with webfakes you can also create that behavior.
+
+Now this means assuming _all_ your tests work with real requests. 
+
+* If a few tests won't work with real requests (say they have a fixture mimicking an [API error](#errors-chapter), or specific answer as if today were a given date) then you can skip these tests when mocking/faking the web service is off. With vcr this means using `vcr::skip_if_vcr_off()`; with httptest and webfakes you'd create your custom skipper.
+* If most tests won't work with real requests, then creating a different folder for tests making real requests makes sense. It might be less unit-y as you could view these tests as integration/contract tests. Maybe they could use [testthat's snapshot testing](https://testthat.r-lib.org/articles/snapshotting.html) (so you could view what's different in the API).
+
+### When to make real requests?
+
+Locally, you might want to make real requests once in a while, in particular before a CRAN release.
+
+On continuous integration you have to learn how to trigger workflows and configure build matrices to e.g.
+
+* Have one build in your build matrix using real requests at each commit (this might be too much, see next section);
+* Have one scheduled workflow once a day or once a week using real requests.
+
+## Why not make only or too many real requests?
+
+The reasons why you can't only make real requests in your tests are the reasons why you are reading these books:
+
+* they are slower;
+* you can't test for API errors;
+* etc.
```

Comment:
will this etc. be filled in later?

Reply:
added a link to the corresponding chapter